ALSA: hda/hdmi: add keep-alive support for ADL-P

kv2019i opened the pull request:
A combination of i915 and ALSA patches to enable keep-alive for display audio (reduces initial playout delay when connected to HDMI/DP receivers).

kv2019i commented:
The i915 part has been sent to DRM tree for review:
https://patchwork.freedesktop.org/series/94551/

kv2019i commented:
@plbossart  @ranj063  @bardliao  @ujfalusi can you do a review on the HDA part? We don't need to merge this to sof-dev and I can send directly to upstream, but while I wait for i915 review feedback, getting some pre-review would be useful.

plbossart commented on sound/pci/hda/patch_hdmi.c:
```diff
@@ -120,6 +120,12 @@ struct hdmi_pcm {
+enum {
+	SILENT_STREAM_OFF = 0,
+	SILENT_STREAM_KAE,
+	SILENT_STREAM_I915
```
a comment on why you have three definitions would help, I wasn't able to follow what that means later in the code. Specifically it's not clear what the _I915 variant might be.

plbossart replied:
or is this legacy with an HDaudio transfer of zeroed data to the iDisp codec?

kv2019i replied:
Thanks @plbossart , exactly the kind of feedback I'm looking for. I'll add comments. Pre-ADL hw has its own keep-alive method that is not standard. We don't actually send zeroed data, but with a special hw sequence we can trigger the i915 to send data to the display link even when audio DMA is not sending anything. This is SILENT_STREAM_I915 and what was added to the codec driver last year.

plbossart commented on sound/pci/hda/patch_hdmi.c:
```diff
@@ -1703,30 +1767,36 @@ static void silent_stream_enable(struct hda_codec *codec,
+	if (spec->silent_stream_type == SILENT_STREAM_I915)
+		silent_stream_enable_i915(codec, per_pin);
+	else
+		silent_stream_enable_kae(codec, per_pin);
```
this is odd, the silent_stream_type can have three values, so you would enable_kae() even when the silent_stream_type is SILENT_STREAM_OFF ??

kv2019i replied:
We don't call this function unless silent-streaming is enabled (i.e. value can never be SILENT_STREAM_OFF if we reach this part of code).

kv2019i replied:
@plbossart  Updated now with hopefully more understandable flow. I also added a piece of text to the Kconfig option to explain the difference.

plbossart commented on sound/pci/hda/patch_hdmi.c:
```diff
@@ -1703,30 +1767,36 @@ static void silent_stream_enable(struct hda_codec *codec,
 	mutex_unlock(&per_pin->lock);
+
+	/* in i915 silent-stream mode, codec must be kept in D0 */
+	if (err || spec->silent_stream_type != SILENT_STREAM_I915)
```
same here, if the type if OFF do you need to do anything?

kv2019i replied:
Ditto here, I tried to make this clearer.

plbossart commented on sound/pci/hda/patch_hdmi.c:
```diff
@@ -1703,30 +1767,36 @@ static void silent_stream_enable(struct hda_codec *codec,
-	int cvt_idx;
+	int cvt_idx, err;
+
+	if (spec->silent_stream_type != SILENT_STREAM_I915) {
```
and here as well?

kv2019i commented:
@plbossart @ranj063 I posted an updated that should be ready for upstream submission. I addressed your comments to the previous version.

However, I cannot submit this yet. The DRM patch needs to be merged before, or otherwise the keepalive function is broken in some kernels (neither new or old silent-stream is fully functional). This will probably go to 5.17 cycle.

kv2019i commented:
i915 part now merged, so time to restart the reviews on this (still can't send to upstream, but time to complete the reviews).

plbossart commented on sound/pci/hda/patch_hdmi.c:
```diff
@@ -1665,18 +1671,76 @@ static void hdmi_present_sense_via_verbs(struct hdmi_spec_per_pin *per_pin,
+	param = snd_hda_codec_read(codec, per_pin->cvt_nid, 0, AC_VERB_GET_DIGI_CONVERT_1, 0);
+	param = (param >> 16) & 0xff;
+	param &= ~AC_DIG3_KAE;
+	snd_hda_codec_write(codec, per_pin->cvt_nid, 0, AC_VERB_SET_DIGI_CONVERT_3, param);
```
nit-pick: in the two functions above, all we do is a read/modify/write, this could be refactored as a single function with an enable parameter.

kv2019i replied:
Thanks, good feedback @plbossart . I must have done something more different in eralier versions. But you are right, they are virtually identical now, let me rework this.

plbossart commented on sound/pci/hda/patch_hdmi.c:
```diff
@@ -3026,6 +3087,27 @@ static int patch_i915_tgl_hdmi(struct hda_codec *codec)
+		 * As the codec can be powered down and keep-alive is
+		 * still sent, there is no need for SILENT_STREAM_I915
+		 * method on these platforms and the "enable_silent_stream"
+		 * module parameter can be ignored.
```
nit-pick: for test purposes it could be interesting to disable this mode - e.g. to check if there's any difference...

kv2019i replied:
@plbossart Indeed. It's a bit of a shame "enable_silent_stream" is a boolean module parameter. For debug it would be also interesting to be able to force old-style silent streaming on new hw. But given this is a bit of a niche feature, I ended up not adding any new options and keeping this simple. If any odd bugs are seen, we can always extend the options (and then we'll have good reasons to do so).

plbossart replied:
sounds good @kv2019i

kv2019i commented:
More changes coming, changing back to draft as this will take a bit of time.

kv2019i commented:
The DRM patch is now in asoc/sound upstream, so we can finally proceed with this. Main changes for this Feb/2022 update:
  - updated the documention,
  - made KAE still disabled by default, requiring opt-in,
  - fixed a bug in disabling the KAE bit
  - added code to apply this to DG2 as well (support added recently to the tree)
 
@jsarha Can you review this as well?

kv2019i commented on sound/pci/hda/Kconfig:
```diff
@@ -285,15 +285,16 @@ config SND_HDA_INTEL_HDMI_SILENT_STREAM
+	  any connected external receiver. This allows to avoid gaps
+	  at start of playback. Many receivers require multiple seconds
+	  to start playing audio after the clock has been stopped.
+	  Power consumption may increase in the receiver connected but no
```
~~Gah, "in the receiver connected" no sense makes. Fix this bug you must!~~ No, change my mind, I'll keep the current wording. Not the best English of my career, but this shall do for now.

jsarha reviewed (approved):
I still did not get to the bottom of this. For instance I could not reverse engineer how the verb system, used for instance in silent_stream_set_kae(), works and what it does there. However, those things I did understand made sense, so this should be good as far as I can tell.

kv2019i commented:
Thank you, sent to upstream directly, closing here -> https://mailman.alsa-project.org/pipermail/alsa-devel/2022-February/195860.html
